src/Workspaces/Core/Portable/Workspace/ProjectSystem/FileWatchedPortableExecutableReferenceFactory.cs

```diff
@@ -37,96 +41,61 @@ internal sealed class FileWatchedPortableExecutableReferenceFactory
     /// <remarks>
     /// This can be removed once https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1843611 is fixed.
     /// </remarks>
-    private readonly ConditionalWeakTable<PortableExecutableReference, string> _previousDisposalLocations = new();
+    private readonly ConditionalWeakTable<TReference, string> _previousDisposalLocations = new();
 
     /// <summary>
     /// <see cref="CancellationTokenSource"/>s for in-flight refreshing of metadata references. When we see a file
     /// change, we wait a bit before trying to actually update the workspace. We need cancellation tokens for those so
     /// we can cancel them either when a flurry of events come in (so we only do the delay after the last modification),
     /// or when we know the project is going away entirely.
     /// </summary>
-    private readonly Dictionary<string, CancellationTokenSource> _metadataReferenceRefreshCancellationTokenSources = [];
+    private readonly Dictionary<string, CancellationTokenSource> _referenceRefreshCancellationTokenSources = [];
 
-    public FileWatchedPortableExecutableReferenceFactory(
-        IFileChangeWatcher fileChangeWatcher)
+    /// <summary>
+    /// Notification when a reference changes on disk.  The string value is the full file path of the changed reference.
+    /// </summary>
+    public event EventHandler<string>? ReferenceChanged;
+
+    public FileWatchedReferenceFactory(IFileChangeWatcher fileChangeWatcher, Func<ImmutableArray<WatchedDirectory>> getAdditionalWatchedDirectories)
     {
         _fileReferenceChangeContext = new Lazy<IFileChangeContext>(() =>
         {
-            var referenceDirectories = new HashSet<string>();
-
-            // On each platform, there is a place that reference assemblies for the framework are installed. These are
-            // rarely going to be changed but are the most common places that we're going to create file watches. Rather
-            // than either creating a huge number of file watchers for every single file, or eventually realizing we
-            // should just watch these directories, we just create the single directory watchers now. We'll collect this
-            // from two places: constructing it from known environment variables, and also for the defaults where those
-            // environment variables would usually point, as a fallback.
-
-            if (Environment.GetEnvironmentVariable("DOTNET_ROOT") is string dotnetRoot && !string.IsNullOrEmpty(dotnetRoot))
-            {
-                referenceDirectories.Add(Path.Combine(dotnetRoot, "packs"));
-            }
-
-            if (PlatformInformation.IsWindows)
-            {
-                referenceDirectories.Add(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86), "Reference Assemblies", "Microsoft", "Framework"));
```

Comment:
this moved into code specific to watching for PEReferences.  Not sure if we need the same code for AnalyzerFileReferences.  @jasonmalinowski i need your thoughts here :)